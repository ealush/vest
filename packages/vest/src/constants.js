/**
 * @type {String} Version number derived from current tag.
 */
export const VERSION = VEST_VERSION;
