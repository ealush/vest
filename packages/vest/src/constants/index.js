export const VERSION = VEST_VERSION;
export const OPERATION_MODE_STATELESS = 'STATELESS';
export const OPERATION_MODE_STATEFUL = 'STATEFUL';
