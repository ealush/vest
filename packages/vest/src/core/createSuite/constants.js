export const SUITE_INIT_ERROR = 'Suite initialization error.';
