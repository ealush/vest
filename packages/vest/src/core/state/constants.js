export const KEY_STATE = '$vestState';
export const KEY_SUITES = '_suites';
export const KEY_CANCELED = '_canceled';
