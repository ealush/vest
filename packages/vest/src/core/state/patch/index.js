import { getSuite } from '..';

/**
 * Updates current suite state with patcher value or output.
 * @param {string} suiteId.
 * @param {Function} patcher Uses to determine next value.
 * @return {Object} Next suite state.
 */
const patch = (suiteId, patcher) => {
  const [state, prevState] = getSuite(suiteId) ?? [];

  const nextState = patcher(state, prevState);

  if (nextState === state) {
    return state;
  }

  getSuite(suiteId)[0] = nextState;
  return nextState;
};

export default patch;
