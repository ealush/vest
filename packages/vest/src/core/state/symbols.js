export const SYMBOL_STATE = Symbol('vestState');
export const SYMBOL_SUITES = Symbol('vestSuites');
export const SYMBOL_CANCELED = Symbol('canceledTests');
