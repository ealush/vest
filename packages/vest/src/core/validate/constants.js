/**
 * @type {String}
 */
export const SUITE_INIT_ERROR = "Suite initialization error.";
