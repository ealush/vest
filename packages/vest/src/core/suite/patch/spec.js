import * as state from '..';
import resetState from '../../../../testUtils/resetState';
import runRegisterSuite from '../../../../testUtils/runRegisterSuite';
import { OPERATION_MODE_STATEFUL } from '../../../constants';
import getSuiteState from '../getSuiteState';
import patch from '.';

const suiteName = 'suite_1';
const suiteId = 'suiteId_1';

let context;

describe('patch', () => {
  let patcher, suite, suiteState, prevState;

  beforeEach(() => {
    resetState();
    context = {
      name: suiteName,
      suiteId,
      operationMode: OPERATION_MODE_STATEFUL,
    };
  });

  describe('When patcher is a function', () => {
    beforeEach(() => {
      patcher = jest.fn(state => ({ ...state, ...{ k: 'v' } }));
      resetState();
      runRegisterSuite(context);
      suite = state.getSuite(suiteId);
      suiteState = suite[0];
      prevState = suite[1];
      patch(suiteId, patcher);
    });
    it('Should set current state value to patcher argument return value', () => {
      expect(state.getSuite(suiteId)[0]).toBe(patcher.mock.results[0].value);
    });
    it('Should pass current and prev state as the patcher arguments', () => {
      expect(patcher).toHaveBeenCalledWith(suiteState, prevState);
    });

    it('Should return next state', () => {
      expect(patch(suiteId, patcher)).toBe(getSuiteState(suiteId));
    });
  });
  describe('When patcher is not a function', () => {
    beforeEach(() => {
      patcher = {};
      resetState();
      runRegisterSuite(context);
    });

    it('Should throw an error', () => {
      expect(() => patch(suiteId, patcher)).toThrow(Error);
    });
  });
});
