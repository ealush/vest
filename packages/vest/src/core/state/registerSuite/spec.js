import { getState } from '..';
import isDeepCopy from '../../../../testUtils/isDeepCopy';
import resetState from '../../../../testUtils/resetState';
import runRegisterSuite from '../../../../testUtils/runRegisterSuite';
import { OPERATION_MODE_STATEFUL } from '../../../constants';
import patch from '../patch';
import { SYMBOL_SUITES } from '../symbols';

const suiteName = 'suite_1';
const suiteId = 'suiteId_1';

let context;
describe('registerSuite', () => {
  let prevSuiteState;

  beforeEach(() => {
    resetState();
    context = {
      name: suiteName,
      suiteId,
      operationMode: OPERATION_MODE_STATEFUL,
    };
  });

  describe('When suite does not exist in state', () => {
    it('Should create initial suite without prevState', () => {
      expect(getState(SYMBOL_SUITES)[suiteId]).toBeUndefined();
      runRegisterSuite(context);
      expect(getState(SYMBOL_SUITES)[suiteId]).toHaveLength(2);
      expect(getState(SYMBOL_SUITES)[suiteId][1]).toBeUndefined();
      expect(getState(SYMBOL_SUITES)[suiteId][0]).toMatchSnapshot();
    });
  });

  describe('When suite exists in state', () => {
    let suite;
    beforeEach(() => {
      runRegisterSuite(context);
      suite = getState(SYMBOL_SUITES)[suiteId];
      prevSuiteState = suite[0];
      runRegisterSuite({ ...context });
    });
    it('Should move prev prevSuiteState into second array cell', () => {
      expect(getState(SYMBOL_SUITES)[suiteId][1]).toBe(prevSuiteState);
    });

    describe('Prev state data handling', () => {
      describe('When lagging and/or pending exist in the previous state', () => {
        let pending, lagging;
        beforeEach(() => {
          pending = [jest.fn(), jest.fn(), jest.fn()];
          lagging = [jest.fn(), jest.fn(), jest.fn()];
          patch(suiteId, state => ({
            ...state,
            pending,
            lagging,
          }));
          runRegisterSuite(context);
        });
        it('Should merge previous pending and lagging into lagging', () => {
          isDeepCopy(suite[0].lagging, [...pending, ...lagging]);
        });

        it('Should match snapshot', () => {
          expect(suite).toMatchSnapshot();
        });
      });
      it('Should nullify prevState pending and lagging', () => {
        expect(suite[1].lagging).toBeNull();
        expect(suite[1].pending).toBeNull();
      });
    });

    it('Should match snapshot', () => {
      expect(suite).toMatchSnapshot();
    });
  });
});
