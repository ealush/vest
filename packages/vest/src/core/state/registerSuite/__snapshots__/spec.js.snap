// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerSuite When suite does not exist in state Should create initial suite without prevState 1`] = `
Object {
  "doneCallbacks": Array [],
  "fieldCallbacks": Object {},
  "groups": Object {},
  "lagging": Array [],
  "name": "suite_1",
  "pending": Array [],
  "suiteId": "suiteId_1",
  "testObjects": Array [],
  "tests": Object {},
}
`;

exports[`registerSuite When suite exists in state Prev state data handling When lagging and/or pending exist in the previous state Should match snapshot 1`] = `
Array [
  Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {},
    "groups": Object {},
    "lagging": Array [
      [MockFunction],
      [MockFunction],
      [MockFunction],
      [MockFunction],
      [MockFunction],
      [MockFunction],
    ],
    "name": "suite_1",
    "pending": Array [],
    "suiteId": "suiteId_1",
    "testObjects": Array [],
    "tests": Object {},
  },
  Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {},
    "groups": Object {},
    "lagging": null,
    "name": "suite_1",
    "pending": null,
    "suiteId": "suiteId_1",
    "testObjects": Array [],
    "tests": Object {},
  },
]
`;

exports[`registerSuite When suite exists in state Should match snapshot 1`] = `
Array [
  Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {},
    "groups": Object {},
    "lagging": Array [],
    "name": "suite_1",
    "pending": Array [],
    "suiteId": "suiteId_1",
    "testObjects": Array [],
    "tests": Object {},
  },
  Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {},
    "groups": Object {},
    "lagging": null,
    "name": "suite_1",
    "pending": null,
    "suiteId": "suiteId_1",
    "testObjects": Array [],
    "tests": Object {},
  },
]
`;
